File: 139/py/139.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        # make dict
        mapD = {}
        for i in range(0, len(wordDict)):
            mapD[wordDict[i]] = 1
        
        # make dp
        n = len(s)
        dp = [False for _ in range(n+1)]
        dp[0] = True

        # state transition
        for i in range(1, n+1):
            for j in range(0, i):
                if dp[j] == True:
                    sub = s[j:i]
                    if mapD.get(sub):
                        dp[i] = True
                        break
        
        return dp[n]

File: 139/go/139.go
// assume dp[i] = if s[i] can be formed by wordDict
// because can't make sure the lrngth of previous one, so assume it dp[j]
// dp[i] = (s[i-j] in wordDict) && dp[j]
// len(dp) = len(s)+1 , set dp[0] = true

func wordBreak(s string, wordDict []string) bool {

    // make a map of wordDict
    mapD := make(map[string]int)

    for i := 0; i < len(wordDict); i++{
        mapD[wordDict[i]] = 1
    }

    // make dp
    n := len(s)
    dp := make([]bool, n+1)

    dp[0] = true

    for i := 1; i <= n; i++{
        for j := 0; j < i; j++{
            if dp[j] == true {
                sub := s[j:i]
                if _, find := mapD[sub]; find{
                    dp[i] = true
                    break
                }
            }
        }
    }
    return dp[n]
}